### ISSUE TITLE
InvalidMemoryAccess for native struct with embedded union

### ISSUE THREAD

**Issue by dgrove-oss** [qbicc/qbicc#1704]
The test program below raises an InvalidMemoryException in the interpreter when compiled on Linux.  I think it is caused by qbicc not liking the union type `epoll_data_t` when it is trying to deal with zero initialization implicit in `auto()`.  It's not clear to me if we should be trying to fix `VarHandleMemory`, etc so that this works, or if we should give up on the union type and just declare `epoll_data_t` as a ptr<void> of the right size and access it via ugly casting...

```
import static org.qbicc.runtime.CNative.*;
import static org.qbicc.runtime.linux.EPoll.*;

public class OffsetTest {
    static final int eventsOffset = eventsOffset();

    public static void main(String[] args) {
        System.out.println(eventsOffset);
    }

    private static int eventsOffset() {
        struct_epoll_event ev = auto();
        return offsetof(ev.events).intValue();
    }
}
```

**Comment by dmlloyd** [comment]
Hmm, I forgot about unions when I wrote `VarHandleMemory`. Let me look at it a bit.

**Comment by dmlloyd** [comment]
Thinking about something along these lines:

```java

**Comment by dgrove-oss** [comment]
That looks plausible.  Wonder if the same pattern could work for some of the horrific unions we have in org.qbicc.runtime.posix.NetIf (`struct_ifreq` and `struct_ifconf`).

**Comment by dmlloyd** [comment]
For conversion idioms I'm thinking about something like this:

```java
        struct_epoll_event ev = auto();
        epoll_data_t.fd data = addr_of(ev.data).sel(epoll_data_t.fd.class);
        data.fd = word(123);
        return data.fd.intValue();
```

**Comment by dmlloyd** [comment]
Right now I'm tinkering to see whether the idiom of storing a dereference (`sel`) in a local variable, and giving it magical alias powers, is a good idea or a terrible idea...

**Comment by dmlloyd** [comment]
OK I'm thinking it's probably not a great idea. Or not a good idea for now, anyway.

This seems like a reasonable (if somewhat verbose) union idiom:

```java
        struct_epoll_event ev = auto();
        addr_of(ev.data).sel(epoll_data_t.fd.class).fd = word(123);
        return addr_of(ev.data).sel(epoll_data_t.fd.class).fd.intValue();
```

**Comment by dmlloyd** [comment]
Fixed by #1750.
